### ISSUE TITLE
Forward hook, Activation holder

### ISSUE THREAD

**Issue by elbaro** [pytorch/pytorch#20637]
## 🚀 Feature

Add `nn.Module.activation` field to hold the last activation (or tuple, etc) returned from forward().

```py
with torch.capture_activations():
    x = net(imgs)

    # last activation from net.conv layer
    loss = cross_entropy(x,y) + L2(net.conv.activation) 

    for (name, activation) in net.named_activations():
        writer.add_histogram('activation_'+name, activation)

# net.conv.activation is released

# capture only desired activations to save memory
with torch.capture_activations(pattern='net/batch_norm_*'):  
    ....
```

## Motivation

- Visualizing activation histogram in tensorboard is not easy.
- Capturing intermediate activations from a network for regularization loss is not easy.

## Pitch

Weight tensors can be enumerated and put into tensorboard like the following:

```py
for (name, param) in model.named_parameters():
    writer.add_histogram('activation_' + name, param, global_step=global_step)
```

However it is not easy to enumerate activations. One needs to insert `writer.add_histogram` in every `forward()` function. For example, capturing intermediate activations of `nn.Sequential` is impossible, and capturing from nn.Module is cumborsome.

```py
def forward(self, x):
    self.memory = []
    x = self.layer1(x)
    self.memory.append(x)
    x = self.layer2(x)
    self.memory.append(x)
    ...
    return x
```

## Alternatives

Add a forward hook with pattern filter. It does not hold the tensor and saves memory for some cases.

**Comment by soumith** [comment]
Thanks for the request.

First, this is pretty impossible in non-jitted pytorch models by design. Unless you are actually intrusively inspecting the Class source code, and parsing `forward` (which can be as complex as one might get), it's simply not doable robustly.

You took a simple case, and showcased it, but it gets pretty complicated. @fmassa and I have thought about this use-case quite extensively in the context of `torchvision`. The complications come when you want intermediate activations that are not simply outputs of layers like nn.Conv2d, but are outputs of some functional part, for example output of `torch.matmul` can be an activation.

The best we have come up so far (after exploring 3 or 4 possibilities) is https://github.com/pytorch/vision/blob/f76e598d47879dbd917bf5936bbd11ff41632787/torchvision/models/_utils.py#L7-L62

I'm closing the feature request because of the above reasons, but I'm happy to discuss the cleanest way one can create a more structured layering system so that you can pull intermediate activations.

@fmassa feel free to share the 3 or 4 possibilities we discussed on this thread.
